## 2. 사이킷런으로 시작하는 머신러닝

### 2.1 사이킷런 소개와 특징

* Scikit-learn : Python ML library 가운데 가장 많이 사용되는 library

* Scikit-learn의 특징 : 교재 87-88쪽을 볼 것.

사이킷런 버전 확인하기

In [1]:
import sklearn
print(sklearn.__version__)

0.24.1


사이킷런 홈페이지 : [Sklearn](https://scikit-learn.org/stable/)

### 2.2 사이킷런을 이용하여 붓꽃(Iris) 데이터 품종 예측하기

- 지도학습 알고리즘으로서의 분류 : 교재 89쪽 설명을 잘 읽어보자.

* Classification(분류) : 붓꽃 꽃잎의 길이, 너비, 꽃받침의 길이, 너비를 가지고 붓꽃의 품종 분류.
 * `features (독립변수)` : sepal length, sepal width, petal length, petal width
 * `label (종속변수)` :  3 different types of irises’ (Setosa, Versicolour, and Virginica)

* `train dataset`, `test dataset` (학습 데이터, 테스트 데이터)
 * 전체 데이터를 학습(훈련) 데이터세트와 테스트 데이터세트 둘로 나눈다.
 * 학습데이터로 머신러닝 모델을 만든 다음 테스트 데이터를 가지고 그 모델의 성능을 평가한다.
 (교재 89쪽 설명을 볼 것).

* 이용할 사이킷런 라이브러리 내의 모듈들 
 * sklearn.datasets : 사이킷런이 제공하는 데이터셋을 만드는 모듈들의 모임 (붓꽃데이터 : load_iris())     
 * `sklearn.tree` : tree에 기반을 둔 머신러닝 알고리즘 모임 (의사결정나무 알고리즘 : DecisionTreeClassifier)     
 * `sklearn.model_selection` : 학습, 검증, 테스트 데이터분리, 최적 hyper parameter로 모델을 평가하는 모듈들의 모임 (학습, 테스트 데이터 분리 : train_test_split) 

* iris data 이외에 사이킷런은 다음과 같은 데이터셋을 제공한다.

 * load_boston: 보스톤 집값 데이터
 * load_diabetes: 당뇨병 환자 데이터
 * load_digits: 손글씨 데이터
 * load_linnerud: multi-output regression 용 데이터
 * load_wine: 와인 데이터
 * load_breast_cancer: 위스콘신 유방암 환자 데이터

**참고 : class**

클래스 : 사이킷런의 분류 알고리즘들을 살펴보기 전에 클래스에 대해 잠깐 공부하자. 

class는 어떤 객체를 만들기 위한 **설계도**에 해당하는 것으로서 알기 쉽게 흔히 '붕어빵틀'에 비유한다. 빵을 만들기 위한 틀은 하나지만 붕어빵장수가 틀에 넣는 재료에 따라 팥빵도 만들고 앙코빵도 만들고 크림빵도 만들 수 있다.

이런 여러빵들처럼 class로부터 만들어진 객체를 instance라고 부른다. 즉 클래스는 객체를 만들기 위한 함수들의 묶음을 의미한다.(잠시 구글 이미지 검생창에 붕어빵 class라고 입력해보자).

함수와 같은 대부분의 객체들이 소문자로 시작하는 이름을 갖는 반면 `클래스의 이름 첫자는` **대문자** 로 시작된다.

어떤 class가 다른 class의 내용을 이어받아서 객체를 만들 수도 있으며, `클래스들 사이의 이런 부모-자식 관게를` **상속(inheritance)** 이라고 부른다.

In [2]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [3]:
import pandas as pd
import numpy as np

# 붓꽃 데이터 세트를 로딩합니다. 
iris = load_iris()

* 붓꽃 데이터 설명

위에서 불러온 iris의 내용을 살펴보자. 

In [4]:
iris

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

In [5]:
type(iris)

sklearn.utils.Bunch

sklearn이 제공하는 데이터셋은 `sklearn.utils.Bunch`라는 자료구조로 되어있는데, 잠시 훑어보면 키와 값으로 된 **딕셔너리** 구조임을 알 수 있다. 그 내용은 크게 `data, target, target_names, feature_names` 넷으로 이루어져 있고 그 사이에 데이터에 대한 이런저런 설명들이 들어있다. 지도학습 머신러닝에서 `target`은 예측해야할 `종속변수`를, `feature`는 `독립변수`를 뜻한다. 

맨 처음에 'data'라는 키 다음에 숫자로 된 배열이 나온다. 그 다음에는 'target'이라는 키, 그리고 0,1,2 숫자로 이루어진 배열이 나온다. 여기서 'target'은 우리가 모형으로 분류하려고 하는 붓꽃의 품종(즉 종속변수)를 나타내고 'data'는 분류 모형에 들어갈 네 가지 독립변수(피처 feature)를 뜻한다. 'target' 아래쪽에는 0,1,2 세 범주가 무엇을 나타내는지 알려주는 'target_names'이 있다. 'setosa'는 0, 'versicolor'는 1, 'virginica'는 2다. 

그리고 맨 아래에는 feature의 이름이 아래와 같이 나온다.

> 'feature_names': 
>
> ['sepal length (cm)',
>  'sepal width (cm)',
>  'petal length (cm)',
>  'petal width (cm)']

* data frame 만들기

먼저 넘파이 배열로 된 feature 데이터만 가져와서 iris_data라고 부르자.

In [6]:
iris_data = iris.data

In [7]:
iris_data.shape

(150, 4)

In [8]:
iris_data[0:5]

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2]])

이어서 종속변수 label이 있는 target의 배열을 `iris_label`이라고 이름붙이자.

In [9]:
# iris.target은 붓꽃 데이터 세트에서 레이블(결정 값) 데이터를 numpy로 가지고 있습니다. 
iris_label = iris.target
print(iris_label)
print()
print(iris.target_names)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]

['setosa' 'versicolor' 'virginica']


다음으로 feature와 target이 들어있는 판다스 데이터프레임을 만들자. 아래 코드는 iris_data 배열 데이터를 먼저 `iris_df`로 만든다(칼럼 이름은 `feature_names`). 이어서 `iris.target`을 lable이라는 칼럼이름을 붙여 iris_df에 추가한다.

In [10]:
# 붓꽃 데이터 세트를 자세히 보기 위해 DataFrame으로 변환합니다. 
iris_df = pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label'] = iris.target
iris_df.head(3)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   

   label  
0      0  
1      0  
2      0

In [11]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     label  
145      2  
146      2  
147      2  
148      2  
149      2

세 종류 분꽃은 각 범주별로 50개씩이다. value_counts() 메서드로 lable 의 빈도분포를 확인 해보자.

In [12]:
iris_df['label'].value_counts()

2    50
1    50
0    50
Name: label, dtype: int64

밑의 np.bincount()로도 가능

In [13]:
np.bincount(iris_df['label'])

array([50, 50, 50], dtype=int64)

- 출석 퀴즈

DataFrame을 정렬할 때 필요한 메서드는? *

In [ ]:
sort_values()

train data, test data를 나누는 함수는? *

In [ ]:
train_test_split()

model 학습을 위한 함수는? *

In [ ]:
fit()

범주형 데이터의 빈도 분포를 알아보는 메서드는? *

In [ ]:
value_counts()

객체(object)를 만들기 위한 설계도를 무엇이라고 부르는가? *

In [ ]:
class

#### 붓꽃 데이터 종류 분류는 다음 네 단계로 이루어진다.
 + 데이터 나누기
 + 학습데이터로 분류 모델 학습
 + 테스트데이터에 대한 예측
 + 모델의 성능 평가

1) `학습, 검증 데이터 나누기`

전체 데이터의 `80%`를 학습데이터로 삼아 분류 모형을 만들고 나머지 `20%`를 가지고 그 모형의 성능을 검증해보자.

**train_test_split()**의 괄호 안에는 먼저 피처 데이터(iris_data)와 레이블 데이터(iris_label)를 입력한다. 이어서 테스트 데이터의 비율을 `test_size=0.2`와 같이 지정하고 난수생성에서 seed와 같은 역할을 하는 `random_state=` 옵션에 아무 숫자나 입력하면 된다.(안넣어도 된다.)

- 중요

In [14]:
X_train, X_test, y_train, y_test = train_test_split(iris_data, iris_label, 
                                                    test_size=0.2, random_state=11)

나눈 결과를 잠시 살펴보자. 150개 데이터 가운데 120개는 학습데이터가 되고 나머지 30개는 테스트 데이터가 되었음을 알 수 있다.

In [15]:
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(120, 4) (30, 4) (120,) (30,)


학습, 테스트 데이터셋에 혹시 세 가지 종류의 붓꽃들이 편중되게 나눠진 것은 아닐까

In [16]:
y_train

array([0, 2, 2, 0, 0, 2, 2, 1, 0, 1, 1, 2, 0, 1, 2, 1, 1, 0, 2, 0, 2, 2,
       1, 2, 1, 0, 0, 1, 0, 0, 2, 2, 2, 0, 0, 0, 1, 0, 1, 2, 2, 1, 1, 2,
       2, 0, 1, 1, 2, 2, 2, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 1, 0, 1, 1, 2,
       1, 0, 0, 0, 1, 1, 1, 2, 1, 0, 1, 2, 0, 2, 2, 1, 0, 0, 0, 2, 1, 0,
       2, 1, 2, 0, 0, 1, 1, 2, 1, 2, 2, 1, 1, 2, 2, 0, 1, 2, 0, 2, 2, 0,
       1, 2, 0, 1, 1, 1, 0, 1, 1, 1])

In [17]:
np.bincount(y_train)

array([41, 40, 39], dtype=int64)

치우침 없이 세 종류가 골고루 들어있음을 알 수 있다.

2) `모델 학습`

의사결정나무 알고리즘을 수행하는 `DecisionTreeClassifier`의 객체를 만든 다음, 붓꽃 `학습 데이터`에 **fit()** 메서드를 적용하여 분류 모델을 만든다. 참고) DecisionTreeClassifier 클래스는 sklearn.tree 모듈 안에 있는데 이미 우리 앞에서 from sklearn,tree import DecisionTreeClassifier 라고 입력하여 이 클래스를 불러두었으므로 아래와 같이 바로 이용하면 된다.

In [18]:
# DecisionTreeClassifier 객체 생성 
dt_clf = DecisionTreeClassifier(random_state=11)

# 학습 수행 
dt_clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=11)

3) `테스트 데이터 예측`

학습 데이터로 만든 모델이 테스트 데이터를 잘 분류하는지 **predict()** 메서드로 알아본다. 

In [19]:
# 학습이 완료된 DecisionTreeClassifier 객체에서 테스트 데이터 세트로 예측 수행. 
pred = dt_clf.predict(X_test)
pred

array([2, 2, 1, 1, 2, 0, 1, 0, 0, 1, 1, 1, 1, 2, 2, 0, 2, 1, 2, 2, 1, 0,
       0, 1, 0, 0, 2, 1, 0, 1])

4) `예측 결과 검증`

`sklearn.metrics` 모듈에 있는 **accuracy_score()**를 이용해서 예측결과(pred)와 실제 값(y_test)가 잘 일치하는지 비교한다. accuracy는 '정확도'를 의미하는데 교재 148쪽에 자세한 설명이 있다.

In [20]:
from sklearn.metrics import accuracy_score
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test, pred)))

예측 정확도: 0.9333


우리가 만든 모형은 30개 가운데 93.3%, 즉 28개를 제대로 예측했다.